File: backend/services/stigmer-server/pkg/domain/agentexecution/filereview/digest.go
// Package filereview is the backend half of the apply-then-review file-change
// subsystem (domain: agentic / agentexecution). It owns the canonical change
// digests, authors FILE_DECIDED events on the append-only file_review ledger,
// and projects FileChangeSet from that ledger. It is the file-review sibling of
// the approval package — the same append-only-ledger discipline instantiated for
// a second lifecycle.
package filereview

import (
	"crypto/sha256"
	"encoding/hex"
	"sort"
	"strings"

	agentexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agentexecution/v1"
)

// FileDigest is the canonical identity+content digest of one captured file
// change — the enforcement key the runner's reconcile checks before applying a
// decision ("what you approve is what gets applied"). It is NEVER a correlation
// key; correlation is by CapturedFileChange.id.
//
// The canonical form digests the content HASHES (before_sha256 / after_sha256),
// not the raw bytes, so the string is pure ASCII and has no line-ending or
// encoding ambiguity to reconcile across Go and Java. kind uses the proto enum
// value name (identical to Java's name()). The format is locked by
// apis/testdata/hitl/file-digest/vectors.json.
func FileDigest(c *agentexecutionv1.CapturedFileChange) string {
	canonical := strings.Join([]string{
		c.GetPathBefore(),
		c.GetPathAfter(),
		c.GetKind().String(),
		c.GetBeforeSha256(),
		c.GetAfterSha256(),
	}, "\n")
	return sha256Hex(canonical)
}

// AggregateDigest is the canonical digest over a change set's whole manifest.
// The per-file digests are sorted so the result is independent of input order (a
// change set is an unordered set of files for identity). The empty set hashes
// the empty string. Enforcement only, never a correlation key.
func AggregateDigest(changes []*agentexecutionv1.CapturedFileChange) string {
	digests := make([]string, 0, len(changes))
	for _, c := range changes {
		digests = append(digests, FileDigest(c))
	}
	sort.Strings(digests)
	return sha256Hex(strings.Join(digests, "\n"))
}

func sha256Hex(s string) string {
	sum := sha256.Sum256([]byte(s))
	return hex.EncodeToString(sum[:])
}

File: backend/services/stigmer-server/src/domain/agentexecution/filereview/digest.ts
/**
 * The canonical file-change digests — ports filereview/digest.go. The
 * filereview module is the backend half of the apply-then-review
 * file-change subsystem: it owns the canonical change digests, authors
 * FILE_DECIDED events on the append-only file_review ledger, and projects
 * FileChangeSet from that ledger — the approval package's append-only
 * discipline instantiated for a second lifecycle.
 *
 * FileDigest is the enforcement key the runner's reconcile checks before
 * applying a decision ("what you approve is what gets applied"); NEVER a
 * correlation key (correlation is by CapturedFileChange.id). The
 * canonical form digests the content HASHES, not raw bytes, so the string
 * is pure ASCII with no encoding ambiguity across editions. kind renders
 * as the proto enum value name (Go .String() / Java name()). The format
 * is locked by apis/testdata/hitl/file-digest/vectors.json.
 */
import { createHash } from "node:crypto";

import { enumToJson } from "@bufbuild/protobuf";

import { FileChangeKindSchema } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/enum_pb";
import type { CapturedFileChange } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/filereview_pb";

export function fileDigest(c: CapturedFileChange): string {
  const canonical = [
    c.pathBefore,
    c.pathAfter,
    enumToJson(FileChangeKindSchema, c.kind) as string,
    c.beforeSha256,
    c.afterSha256,
  ].join("\n");
  return sha256Hex(canonical);
}

/**
 * The canonical digest over a change set's whole manifest. Per-file
 * digests are sorted so the result is independent of input order (a
 * change set is an unordered set of files for identity). The empty set
 * hashes the empty string. Enforcement only, never a correlation key.
 */
export function aggregateDigest(changes: CapturedFileChange[]): string {
  const digests = changes.map(fileDigest);
  digests.sort();
  return sha256Hex(digests.join("\n"));
}

function sha256Hex(s: string): string {
  return createHash("sha256").update(s, "utf-8").digest("hex");
}

/**
 * The raw-bytes content address CAS blobs are keyed by (Node
 * createHash("sha256").digest("hex") — lowercase hex, exactly Go's
 * hex.EncodeToString).
 */
export function sha256HexBytes(b: Uint8Array): string {
  return createHash("sha256").update(b).digest("hex");
}
